### PR TITLE
Piper/cleanup linting recursion and mutability

### PR DESCRIPTION
**Pull request by pipermerriam**
### What this does

This fixes 3 things at once.

- ensure recursion limit setting doesn't reduce recursion limit
- convert `FQ` classes to use immutable data structures
- linting issues

### CODE REVIEW

File: py_ecc/optimized_bls12_381/optimized_field_elements.py [hunk 1]
```diff
@@ -36,7 +28,7 @@ def __init__(self, n):
             self.n = n.n
         else:
             self.n = n % field_modulus
-        assert isinstance(self.n, int_types)
+        assert isinstance(self.n, int)
```
Comment (carver) [inline py_ecc/optimized_bls12_381/optimized_field_elements.py]: Not sure it's worth double-checking, but thought I would add the reminder that `isinstance(bool(), int) == True`

**Comment by carver** [comment]
Oh I see py2 support was already dropped, this just removes some cruft.

**Comment by Bhargavasomu** [comment]
@pipermerriam could you merge this so that I can rebase on top of this, for closing #29?

**Comment by pipermerriam** [comment]
@Bhargavasomu done.

**Pull request merged**